src/morale.cpp

```diff
@@ -137,10 +218,379 @@ void morale_point::proceed( int ticks )
     }
 
     age += ticks;
+}
+
+void player_morale::add( morale_type type, int bonus, int max_bonus,
+                         int duration, int decay_start,
+                         bool capped, const itype *item_type )
+{
+    // Search for a matching morale entry.
+    for( auto &m : points ) {
+        if( m.get_type() == type && m.get_item_type() == item_type ) {
+            const int prev_bonus = m.get_net_bonus();
+
+            m.add( bonus, max_bonus, duration, decay_start, capped );
+            if( m.get_net_bonus() != prev_bonus ) {
+                invalidate();
+            }
+            return;
+        }
+    }
+
+    morale_point new_morale( type, item_type, bonus, duration, decay_start );
+
+    if( !new_morale.is_expired() ) {
+        points.push_back( new_morale );
+        invalidate();
+    }
+}
+
+int player_morale::has( morale_type type, const itype *item_type ) const
+{
+    for( auto &m : points ) {
+        if( m.get_type() == type && ( item_type == nullptr || m.get_item_type() == item_type ) ) {
+            return m.get_net_bonus();
+        }
+    }
+    return 0;
+}
+
+void player_morale::remove( morale_type type, const itype *item_type )
+{
+    for( size_t i = 0; i < points.size(); ++i ) {
+        if( points[i].get_type() == type && points[i].get_item_type() == item_type ) {
+            if( points[i].get_net_bonus() > 0 ) {
+                invalidate();
+            }
+            points.erase( points.begin() + i );
+            break;
+        }
+    }
+}
+
+morale_mult player_morale::get_traits_mult( const player &p ) const
```

Comment:
Instead of passing a player here, the player_morale class can cache whether it has these bonuses, and the player can tell the player_morale class whenever it gains or loses a trait (all of them, not just the ones the player_morale class cares about).
This is an example of the [observer pattern](http://gameprogrammingpatterns.com/observer.html), except with only one listener, eventually we might make a full-blown observer class to help out with this kind of thing.